packages/driver/src/cy/commands/querying/querying.ts

```
@@ -591,216 +229,172 @@ export default (Commands, Cypress, cy, state) => {
-          withinSubject: subject || state('withinSubject') || cy.$$('body'),
-          filter: true,
-          log: false,
-          // retry: false ## dont retry because we perform our own element validation
-          verify: false, // dont verify upcoming assertions, we do that ourselves
-        })
-
-        return cy.now('get', selector, getOptions).then(($el) => {
-          if ($el && $el.length) {
-            $el = $dom.getFirstDeepestElement($el)
+      log && log.set({
+        message: $utils.stringify(_.compact([filter, text])),
+        $el,
+        type: subject.is('body') ? 'parent' : 'child',
+        consoleProps: () => {
+          return {
+            Content: text,
+            'Applied To': $dom.getElements(subject),
+            Yielded: $el.get(0),
+            Elements: $el.length,
           }
+        },
+      })
 
-          setEl($el)
-
-          return cy.verifyUpcomingAssertions($el, options, {
-            onRetry: resolveElements,
-            onFail (err) {
-              switch (err.type) {
-                case 'length':
-                  if (err.expected > 1) {
-                    return $errUtils.throwErrByPath('contains.length_option', { onFail: options._log })
-                  }
-
-                  break
-                case 'existence':
-                  return err.message = getErr(err)
-                default:
-                  break
-              }
-
-              return null
-            },
-          })
-        })
-      }
-
-      return Promise
-      .try(resolveElements)
-    },
+      return $el
+    }
   })
 
-  Commands.add('shadow', { prevSubject: 'element' }, (subject, options) => {
-    const userOptions = options || {}
-
-    options = _.defaults({}, userOptions, { log: true })
-
-    const consoleProps: Record<string, any> = {
-      'Applied To': $dom.getElements(subject),
-    }
+  Commands._addQuery('shadow', function contains (userOptions: ShadowOptions = {}) {
```

The way `_addQuery` is currently implemented, this changes .shadow from being a child command which requires an element subject to a dual command which an optional previous subject. This is new behavior that I don't anticipate is intentinal.

---

Seems like this is handled below by `cy.ensureSubjectByType(subject, 'element')` instead of being part of the `_addQuery` functionality. Is that right?

---

Aye. _addQuery does not have builtin functionality to assert on how queries can be used - each query that wants to has to explicitly run its own `ensure()`s to validate its position and incoming subject.

The reasons for this are a bit multifaceted, but it boils down to "parent / child / dual don't map cleanly to how Cypress implements them." You can execute parent commands after child commands, as long as the parent command returns null. And for commands we already have code to *derive* parent/child/dual from the `prevSubject` anyway. `prevSubject` is the "natural" concept, and parent/child/dual is a computed property.

So for Queries, I skipped that whole set of logic. The goal is to make it clearer when and how this code executes, to someone reading or writing queries. Each query states, explicitly with an `ensures` statement (rather than via an argument to _addQuery), exactly what subjects are allowed.

---

@BlueWinds Then why does this default to `type: 'dual'` and `previousSubject: 'optional'`? https://github.com/cypress-io/cypress/blob/issue-7306-addQuery-custom-commands/packages/driver/src/cypress/cy.ts#L85

---

Shouldn't it default to `parent` and `none`  and the query commands that change this behavior override this?

---

It defaults to dual/optional because I want to bypass the existing logic entirely. This goes pretty deep into the weeds of how `prevSubject` is implemented.

`prevSubject` is checked *before* invoking the command / query. But for queries, we instead want to throw an error *inside the retry loop* - the subject might (is expected to!) change between retries, and we want to run the validation each time.

So we use 'optional' to signal that Cypress shouldn't try running any assertions before invoking the command - any incoming subject is acceptable. The query will validate the incoming subject later, in the inner queryFn.

---

Can you link through where this is? I have been pretty deep in this code recently, and my understanding of addCommand, is that we are actually [wrapping the commands callback ](https://github.com/cypress-io/cypress/blob/issue-7306-addQuery-custom-commands/packages/driver/src/cypress/cy.ts#L744)(which we aren't doing with _addQuery()) which enforces the previous subject check for us.

---

Ok, went through the code and looks like my explanation is out of date - what I said is why I *originally* had optional / dual, but it no longer applies. Thank you for pressing on the matter!

`optional` and `dual` aren't needed at all, so I've removed them. Queries just plain old don't interact with the concept of `prevSubject` / `type` anymore - they follow their own codepath which bypasses them entirely, and my above explanation was badly out of date.